### PR TITLE
Encode logfile option with URLEncoder

### LINKED ISSUE #786: Log file can not be found for multiproject Gradle projects
### Versions used 

Akka version: 
akka-grpc-gradle-plugin:0.7.3

### Expected Behavior

When building a Gradle project with multiple subprojects (e.g., a root gradle build and at least one subproject build), building the project from the root project should work.

Specifically: `./gradlew build` from the root project's directory should complete successfully.

The same goes for if the project is built not from the root project, but from the specific subproject.

### Actual Behavior

If the build is triggered from the **_subproject_**, things work as expected for that subproject.

If the build is triggered from the **_root_**, the Gradle plugin complains that a log does not exist (see below for relevant logs).

This regression is due to a change made in PR #674, which addresses issue #669.

### Relevant logs

```
> Task :subProject:generateProto FAILED

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':subProject:generateProto'.
> protoc: stdout: . stderr: /Users/eorendain/Documents/myProject/subProject/src/main/protobuf: warning: directory does not exist.
  /Users/eorendain/Documents/myProject/subProject/app/protobuf: warning: directory does not exist.
  /Users/eorendain/Documents/myProject/subProject/app/proto: warning: directory does not exist.
  bytestream.proto: warning: Import google/protobuf/wrappers.proto but not used.
  Exception in thread "main" java.io.FileNotFoundException: build/akka-grpc-gradle-plugin.log (No such file or directory)
        at java.base/java.io.FileOutputStream.open0(Native Method)
        at java.base/java.io.FileOutputStream.open(FileOutputStream.java:298)
        at java.base/java.io.FileOutputStream.<init>(FileOutputStream.java:237)
        at java.base/java.io.FileOutputStream.<init>(FileOutputStream.java:187)
        at java.base/java.io.PrintWriter.<init>(PrintWriter.java:208)
        at java.base/java.io.PrintWriter.<init>(PrintWriter.java:248)
        at akka.grpc.gen.FileLogger.<init>(Logging.scala:35)
        at akka.grpc.gen.Main$.$anonfun$logger$1(Main.scala:48)
        at scala.Option.map(Option.scala:230)
        at akka.grpc.gen.Main$.delayedEndpoint$akka$grpc$gen$Main$1(Main.scala:48)
        at akka.grpc.gen.Main$delayedInit$body.apply(Main.scala:14)
        at scala.Function0.apply$mcV$sp(Function0.scala:39)
        at scala.Function0.apply$mcV$sp$(Function0.scala:39)
        at scala.runtime.AbstractFunction0.apply$mcV$sp(AbstractFunction0.scala:17)
        at scala.App.$anonfun$main$1$adapted(App.scala:80)
        at scala.collection.immutable.List.foreach(List.scala:392)
        at scala.App.main(App.scala:80)
        at scala.App.main$(App.scala:78)
        at akka.grpc.gen.Main$.main(Main.scala:14)
        at akka.grpc.gen.Main.main(Main.scala)
  --akkaGrpc_out: protoc-gen-akkaGrpc: Plugin failed with status code 1.
```

### Reproducible Test Case

Apologies, I don't have much time to whip up a test case, but one would be simple to create.  Simply create a multi-project Gradle build wherein a subproject uses this Akka gRPC Gradle plugin, and run the commands as outlined above.

### Fix

I'll submit a PR with a fix in a few moments for testing/feedback.

### PR DESCRIPTION
**Pull request by jasonxh**
References #786

## Issue
Currently, applying the gradle plugin to a subproject will fail gradle build from root project dir, due to a mismatch in the expectation of where the `logfile` is located. The gradle plugin calculates a relative path to the subproject, while the protoc codegen plugin evaluates it as relative to the current working dir (which is root dir).

## Background
There had been several back-and-forths regarding how to pass the `logfile` option in the past, ranging from one-off character substitution, to relative paths, to absolute paths (see #674 and #787). The root of the problem is that protoc codegen plugin options are passed in as a single string, encoded in a special format like this:

```
--akkaGrpc_out=<key1>=<value1>,<key2>=<value2>:<out_dir>
```

 It's easy to see how several characters play a special role in this encoding, and thus should not appear in keys or values: `[=,:]`.

This problem was originally surfaced as a Windows-only problem, where absolute file paths include `C:\` (see #669). However, it doesn't actually matter if we are passing `logfile` as absolute or relative, since those special characters mentioned above can creep in either way. Granted, no sane person would name their files like that, but it's possible.

## Proposed fix
Encode `logfile` with [URLEncoder](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/net/URLEncoder.html) when composing options, and decode it with [URLDecoder](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/net/URLDecoder.html) inside the protoc codegen plugin. This is a well-known encoding scheme that allows only `[a-zA-Z0-9.-*_+%]` in the encoded output. Note that none of the special characters mentioned above are allowed here.

I also renamed the option to `logfile_enc` to make it an explicit breaking change, in case the codegen plugin is being used in other context. But I don't feel strongly about this, so feel free to suggest otherwise.

## Testing
Verified to fix #786 on my Mac. Would be great if someone can help test this on a Windows machine.

### CODE REVIEW

**Comment by lightbend-cla-validator** [comment]
Hi @jasonxh,

Thank you for your contribution! We really value the time you've taken to put this together.

Before we proceed with reviewing this pull request, please sign the Lightbend Contributors License Agreement:

[https://www.lightbend.com/contribute/cla](https://www.lightbend.com/contribute/cla)

**Comment by raboof** [comment]
OK TO TEST

**Comment by raboof** [comment]
@hithran @orendain @eshepelyuk thoughts?

**Comment by eshepelyuk** [comment]
> @hithran @orendain @eshepelyuk thoughts?
looks ok to me.

File: codegen/src/main/scala/akka/grpc/gen/Main.scala [hunk 1]
```diff
@@ -45,7 +46,8 @@ object Main extends App {
   private val extraGenerators: List[String] =
     parameters.getOrElse("extra_generators", "").split(";").toList.filter(_ != "")
 
-  private val logger = parameters.get("logfile").map(new FileLogger(_)).getOrElse(SilencedLogger)
+  private val logger: Logger =
+    parameters.get("logfile_enc").map(URLDecoder.decode(_, "utf-8")).map(new FileLogger(_)).getOrElse(SilencedLogger)
```
Comment (raboof) [inline codegen/src/main/scala/akka/grpc/gen/Main.scala:50]: can we fall back to the `logfile` parameter when `logfile_enc` is not present?

Comment (jasonxh) [reply codegen/src/main/scala/akka/grpc/gen/Main.scala]: Added fallback logic

**Review by raboof — APPROVED** [review]
Nice, thanks!

**Comment by hithran** [comment]
> @hithran @orendain @eshepelyuk thoughts?

@jasonxh Great solution! Thanks!

**Pull request merged**